<a href="https://colab.research.google.com/github/JustinRizzo/Team_DN5_Assignment2/blob/main/teamDN5_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Explore the Data

First connecting to GCP

In [6]:
from google.colab import auth
import os

# Authenticate your Google account
auth.authenticate_user()

# --- CONFIGURATION ---
# NOTE: Verify this is your correct Google Cloud Project ID
PROJECT_ID = "mgmt599-justinrizzo-lab2"
REGION = "us-central1"

# Use the bucket from your screenshot
GCS_BUCKET = "mgmt599-justinrizzo-assignment2"

# --- Resource Paths ---
# Define the full paths for our GCS and BigQuery resources

# Input file updated to point to train.csv
input_file = f"gs://{GCS_BUCKET}/kaggle-store-sales/train.csv"

# Temporary location for Dataflow jobs, using the 'temp' folder in your bucket
temp_location = f"gs://{GCS_BUCKET}/temp/"

# BigQuery destination details
BIGQUERY_DATASET = "kaggle_store_sales"
# A new table name for the training data
BIGQUERY_TABLE = "train"
table_spec = f"{PROJECT_ID}:{BIGQUERY_DATASET}.{BIGQUERY_TABLE}"

# Define the BigQuery table schema based on your train.csv file
table_schema = "id:INTEGER,date:DATE,store_nbr:INTEGER,family:STRING,sales:FLOAT,onpromotion:INTEGER"

print("✅ Configuration is complete and updated for train.csv.")
print(f"Input file: {input_file}")
print(f"Table spec: {table_spec}")

✅ Configuration is complete and updated for train.csv.
Input file: gs://mgmt599-justinrizzo-assignment2/kaggle-store-sales/train.csv
Table spec: mgmt599-justinrizzo-lab2:kaggle_store_sales.train


# Task
Explore the dataset "gs://cloud-samples-data/ai-platform/customers_transactions/train.csv" to answer the following questions: How many stores are in the dataset? What date range does it cover? What are the product families?

## Load the dataset

### Subtask:
Load the `train.csv` file from the specified GCS bucket into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the data from the specified GCS path into a DataFrame.



In [7]:
import pandas as pd

# Define the input file path for the train data
train_input_file = "gs://cloud-samples-data/ai-platform/customers_transactions/train.csv"

# Read the CSV file into a pandas DataFrame
df_train = pd.read_csv(train_input_file)

# Display the first few rows of the DataFrame to verify it loaded correctly
display(df_train.head())

FileNotFoundError: b/cloud-samples-data/o/ai-platform%2Fcustomers_transactions%2Ftrain.csv

**Reasoning**:
The previous attempt to load the data failed because the file was not found at the specified path. I need to use the correct GCS path for the `train.csv` file, which was defined in the configuration section.



In [8]:
# Use the correct input file path from the configuration
df_train = pd.read_csv(input_file)

# Display the first few rows of the DataFrame to verify it loaded correctly
display(df_train.head())

id        date  store_nbr      family  sales  onpromotion
0   0  2013-01-01          1  AUTOMOTIVE    0.0            0
1   1  2013-01-01          1   BABY CARE    0.0            0
2   2  2013-01-01          1      BEAUTY    0.0            0
3   3  2013-01-01          1   BEVERAGES    0.0            0
4   4  2013-01-01          1       BOOKS    0.0            0

## Explore the data

### Subtask:
Display the first few rows and the column information of the DataFrame to understand its structure and content.


**Reasoning**:
Display the first few rows and the column information of the DataFrame to understand its structure and content as requested by the instructions.



In [9]:
display(df_train.head())
df_train.info()

id        date  store_nbr      family  sales  onpromotion
0   0  2013-01-01          1  AUTOMOTIVE    0.0            0
1   1  2013-01-01          1   BABY CARE    0.0            0
2   2  2013-01-01          1      BEAUTY    0.0            0
3   3  2013-01-01          1   BEVERAGES    0.0            0
4   4  2013-01-01          1       BOOKS    0.0            0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000888 entries, 0 to 3000887
Data columns (total 6 columns):
 #   Column       Dtype  
---  ------       -----  
 0   id           int64  
 1   date         object 
 2   store_nbr    int64  
 3   family       object 
 4   sales        float64
 5   onpromotion  int64  
dtypes: float64(1), int64(3), object(2)
memory usage: 137.4+ MB


## Answer the questions

### Subtask:
Determine the number of unique stores, identify the date range, and identify the unique product families in the dataset.


**Reasoning**:
Calculate the number of unique stores, determine the date range, and identify the unique product families from the dataframe.



In [10]:
# 1. Calculate and print the number of unique values in the 'store_nbr' column
num_unique_stores = df_train['store_nbr'].nunique()
print(f"Number of unique stores: {num_unique_stores}")

# 2. Convert the 'date' column to datetime objects
df_train['date'] = pd.to_datetime(df_train['date'])

# 3. Find and print the minimum and maximum dates
min_date = df_train['date'].min()
max_date = df_train['date'].max()
print(f"Date range: {min_date} to {max_date}")

# 4. Find and print all unique values in the 'family' column
unique_families = df_train['family'].unique()
print("Unique product families:")
for family in unique_families:
    print(family)

Number of unique stores: 54
Date range: 2013-01-01 00:00:00 to 2017-08-15 00:00:00
Unique product families:
AUTOMOTIVE
BABY CARE
BEAUTY
BEVERAGES
BOOKS
BREAD/BAKERY
CELEBRATION
CLEANING
DAIRY
DELI
EGGS
FROZEN FOODS
GROCERY I
GROCERY II
HARDWARE
HOME AND KITCHEN I
HOME AND KITCHEN II
HOME APPLIANCES
HOME CARE
LADIESWEAR
LAWN AND GARDEN
LINGERIE
LIQUOR,WINE,BEER
MAGAZINES
MEATS
PERSONAL CARE
PET SUPPLIES
PLAYERS AND ELECTRONICS
POULTRY
PREPARED FOODS
PRODUCE
SCHOOL AND OFFICE SUPPLIES
SEAFOOD


## Present the findings

### Subtask:
Summarize the answers to the questions.


**Reasoning**:
Summarize the answers to the questions using the variables computed in the previous step.



In [11]:
print(f"Summary of the dataset:")
print(f"- Number of stores in the dataset: {num_unique_stores}")
print(f"- Date range covered: {min_date.strftime('%Y-%m-%d')} to {max_date.strftime('%Y-%m-%d')}")
print(f"- Unique product families: {', '.join(unique_families)}")

Summary of the dataset:
- Number of stores in the dataset: 54
- Date range covered: 2013-01-01 to 2017-08-15
- Unique product families: AUTOMOTIVE, BABY CARE, BEAUTY, BEVERAGES, BOOKS, BREAD/BAKERY, CELEBRATION, CLEANING, DAIRY, DELI, EGGS, FROZEN FOODS, GROCERY I, GROCERY II, HARDWARE, HOME AND KITCHEN I, HOME AND KITCHEN II, HOME APPLIANCES, HOME CARE, LADIESWEAR, LAWN AND GARDEN, LINGERIE, LIQUOR,WINE,BEER, MAGAZINES, MEATS, PERSONAL CARE, PET SUPPLIES, PLAYERS AND ELECTRONICS, POULTRY, PREPARED FOODS, PRODUCE, SCHOOL AND OFFICE SUPPLIES, SEAFOOD


## Summary:

### Q&A

*   **How many stores are in the dataset?**
    There are 54 unique stores in the dataset.
*   **What date range does it cover?**
    The dataset covers the date range from 2013-01-01 to 2017-08-15.
*   **What are the product families?**
    There are 33 unique product families, including AUTOMOTIVE, BABY CARE, BEAUTY, BEVERAGES, BOOKS, BREAD/BAKERY, CELEBRATION, CLEANING, DAIRY, DELI, EGGS, FROZEN FOODS, GROCERY I, GROCERY II, HARDWARE, HOME AND KITCHEN I, HOME AND KITCHEN II, HOME APPLIANCES, HOME CARE, LADIESWEAR, LAWN AND GARDEN, LINGERIE, LIQUOR,WINE,BEER, MAGAZINES, MEATS, PERSONAL CARE, PET SUPPLIES, PLAYERS AND ELECTRONICS, POULTRY, PREPARED FOODS, PRODUCE, SCHOOL AND OFFICE SUPPLIES, and SEAFOOD.

### Data Analysis Key Findings

*   The dataset contains 3,000,888 entries and 6 columns (`id`, `date`, `store_nbr`, `family`, `sales`, and `onpromotion`).
*   All columns have no missing values.
*   The dataset includes data for 54 distinct stores.
*   The data spans from January 1, 2013, to August 15, 2017.
*   There are 33 different product families recorded in the dataset.

### Insights or Next Steps

*   The dataset provides a comprehensive view of sales transactions across multiple stores and product families over a significant period, suitable for time-series analysis or sales forecasting.
*   Further analysis could explore sales trends by store, product family, or over time, and investigate the impact of promotions (`onpromotion` column) on sales.


Purpose: Before we can build our pipeline, we need to install the necessary Python libraries. We will install apache-beam with the [gcp] extra, which includes all the components needed to run our pipeline on the Google Cloud Dataflow service.

Important Note on Dependencies: Google Colab comes with many pre-installed packages. Sometimes, these can conflict with the specific versions required by new libraries. In this case, apache-beam requires a newer version of a package called dill. To solve this, we will install both apache-beam and the required version of dill in a single command, which helps the package manager resolve the conflict correctly.

Action Required:

Run this cell to install the libraries.

After the installation completes, you must restart the Colab runtime. This ensures that the newly installed package versions are loaded correctly. Go to the menu and select Runtime > Restart runtime.

In [12]:
# --- Install Apache Beam and its dependencies in a single command ---
# This command installs the Apache Beam SDK with Google Cloud Platform support
# and simultaneously ensures that the 'dill' package is at version 0.3.7 or newer,
# which resolves a common dependency conflict in the Colab environment.

!pip install --upgrade pip
!pip install --upgrade dill
!pip install --upgrade google-cloud-bigquery
!pip install --quiet apache-beam[gcp]

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 22.0 MB/s eta 0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 24.1.2
    Uninstalling pip-24.1.2:
      Successfully uninstalled pip-24.1.2
  Attempting uninstall: dill
    Found existing installation: dill 0.3.8
    Uninstalling dill-0.3.8:
      Successfully uninstalled dill-0.3.8
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datasets 4.0.0 requires dill<0.3.9,>=0.3.0, but you have dill 0.4.0 which is incompatible.
  Attempting uninstall: google-cloud-bigquery
    Found existing installation: google-cloud-bigquery 3.35.0
    Uninstalling google-cloud-bigquery-3.35.0:
      Successfully uninstalled google-cloud-bigquery-3.35.0


  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.4/17.4 MB 122.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 76.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.9/527.9 kB 20.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 113.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 52.9 MB/s eta 0:00:00
  DEPRECATION: Building 'crcmod' using the legacy setup.py bdist_wheel mechanism, which will be removed in a future version. pip 25.3 will enforce this behaviour change. A possible replacement is to use the standardized build interface by setting the `--use-pep517` option, (possibly combined with `--no-build-isolation`), or adding a `pyproject.toml` file to the source tree of 'crcmod'. Discussion ca

#Pipeline for train dataset

In [17]:
%%writefile pipeline_sales.py
import apache_beam as beam
from apache_beam.options.pipeline_options import PipelineOptions
import logging

# --- CONFIGURATION ---
# !!! IMPORTANT: SET YOUR TEAM NUMBER HERE !!!
# TODO: Replace "X" with your actual team number
TEAM_NUMBER = "DN5" # e.g., "1", "2", etc.

PROJECT_ID = "mgmt599-justinrizzo-lab2"
GCS_BUCKET = "mgmt599-justinrizzo-assignment2"
BIGQUERY_DATASET = f"store_sales_team_{TEAM_NUMBER}"
BIGQUERY_TABLE = "sales_data"
REGION = "us-central1"

# --- Resource Paths ---
INPUT_FILE = f"gs://{GCS_BUCKET}/kaggle-store-sales/train.csv"
TABLE_SPEC = f"{PROJECT_ID}:{BIGQUERY_DATASET}.{BIGQUERY_TABLE}"
TEMP_LOCATION = f"gs://{GCS_BUCKET}/temp/"
STAGING_LOCATION = f"gs://{GCS_BUCKET}/staging/"

# --- DATA TRANSFORMATION FUNCTION ---
def parse_train_csv(line):
    """
    Parses a single line from train.csv and returns a dictionary.
    Handles type conversions and skips the header row.
    Includes error handling for malformed rows.
    """
    header = "id,date,store_nbr,family,sales,onpromotion"
    if line.strip().lower() == header.lower():
        logging.info("Skipping header row.")
        return []  # Skip header

    fields = line.split(',')
    # Ensure the correct number of fields before attempting to parse
    if len(fields) != 6:
        logging.warning(f"Skipping malformed row (incorrect number of fields): {line}")
        return []

    try:
        # Create a dictionary for each row, converting types as needed
        # Added explicit error logging for type conversion issues
        return [{
            "id": int(fields[0]),
            "date": fields[1],  # Date is already in 'YYYY-MM-DD' format, BigQuery will handle this with DATE type
            "store_nbr": int(fields[2]),
            "family": fields[3],
            "sales": float(fields[4]),
            "onpromotion": int(fields[5])
        }]
    except (ValueError, IndexError) as e:
        # Log rows that don't match the expected format or type conversions
        logging.warning(f"Skipping malformed row (type conversion error): {line} | Error: {e}")
        return []

def run():
    """Defines and runs the Beam pipeline."""
    # --- PIPELINE OPTIONS ---
    options = PipelineOptions(
        runner='DataflowRunner',
        project=PROJECT_ID,
        job_name=f'gcs-to-bq-sales-data-team{TEAM_NUMBER.lower()}',
        staging_location=STAGING_LOCATION,
        temp_location=TEMP_LOCATION,
        region=REGION,
        # Add save_main_session to make defined functions available to workers
        save_main_session=True
    )

    # BigQuery table schema
    # Ensure the schema matches the expected data types and column names
    table_schema = "id:INTEGER,date:DATE,store_nbr:INTEGER,family:STRING,sales:FLOAT,onpromotion:INTEGER"

    # Corrected: Access job_name using get_all_options()
    logging.info(f"Starting Dataflow job: {options.get_all_options()['job_name']}")
    logging.info(f"Reading from: {INPUT_FILE}")
    logging.info(f"Writing to BigQuery table: {TABLE_SPEC}")
    logging.info(f"BigQuery schema: {table_schema}")


    # Define the pipeline
    with beam.Pipeline(options=options) as p:
        (
            p
            | 'ReadSalesData' >> beam.io.ReadFromText(INPUT_FILE, skip_header_lines=1) # Use skip_header_lines instead of manual skipping
            | 'ParseSalesCSV' >> beam.FlatMap(parse_train_csv)
            | 'WriteSalesToBigQuery' >> beam.io.WriteToBigQuery(
                table=TABLE_SPEC,
                schema=table_schema,
                create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
                write_disposition=beam.io.BigQueryDisposition.WRITE_TRUNCATE
            )
        )
    logging.info(f"Job '{options.get_all_options()['job_name']}' submitted successfully.")

if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    run()

Overwriting pipeline_sales.py


#Launch the job

In [18]:
!python pipeline_sales.py

INFO:root:Starting Dataflow job: gcs-to-bq-sales-data-teamdn5
INFO:root:Reading from: gs://mgmt599-justinrizzo-assignment2/kaggle-store-sales/train.csv
INFO:root:Writing to BigQuery table: mgmt599-justinrizzo-lab2:store_sales_team_DN5.sales_data
INFO:root:BigQuery schema: id:INTEGER,date:DATE,store_nbr:INTEGER,family:STRING,sales:FLOAT,onpromotion:INTEGER
INFO:root:Runner defaulting to pickling library: cloudpickle.
INFO:apache_beam.runners.dataflow.dataflow_runner:Pipeline has additional dependencies to be installed in SDK worker container, consider using the SDK container image pre-building workflow to avoid repetitive installations. Learn more on https://cloud.google.com/dataflow/docs/guides/using-custom-containers#prebuild
INFO:apache_beam.runners.dataflow.internal.apiclient:Starting GCS upload to gs://mgmt599-justinrizzo-assignment2/staging/gcs-to-bq-sales-data-teamdn5.1753591684.264621/submission_environment_dependencies.txt...
INFO:apache_beam.runners.dataflow.internal.apiclient

#Pipeline for stores dataset

In [19]:
%%writefile pipeline_stores.py
import apache_beam as beam
from apache_beam.options.pipeline_options import PipelineOptions
import logging

# --- CONFIGURATION ---
# !!! IMPORTANT: SET YOUR TEAM NUMBER HERE !!!
# TODO: Replace "X" with your actual team number
TEAM_NUMBER = "DN5" # e.g., "1", "2", etc.

PROJECT_ID = "mgmt599-justinrizzo-lab2"
GCS_BUCKET = "mgmt599-justinrizzo-assignment2"
BIGQUERY_DATASET = f"store_sales_team_{TEAM_NUMBER}"
BIGQUERY_TABLE = "store_info"
REGION = "us-central1"

# --- Resource Paths ---
INPUT_FILE = f"gs://{GCS_BUCKET}/kaggle-store-sales/stores.csv"
TABLE_SPEC = f"{PROJECT_ID}:{BIGQUERY_DATASET}.{BIGQUERY_TABLE}"
TEMP_LOCATION = f"gs://{GCS_BUCKET}/temp/"
STAGING_LOCATION = f"gs://{GCS_BUCKET}/staging/"

# --- DATA TRANSFORMATION FUNCTION ---
def parse_stores_csv(line):
    """
    Parses a single line from stores.csv and returns a dictionary.
    Includes error handling for malformed rows.
    """
    # The header is checked and skipped by the ReadFromText transform
    fields = line.split(',')
    # Ensure the correct number of fields before attempting to parse
    if len(fields) != 5:
        logging.warning(f"Skipping malformed row (incorrect number of fields): {line}")
        return []

    try:
        # Create a dictionary for each row, converting types as needed
        return [{
            "store_nbr": int(fields[0]),
            "city": fields[1],
            "state": fields[2],
            "type": fields[3],
            "cluster": int(fields[4])
        }]
    except (ValueError, IndexError) as e:
        # Log rows that don't match the expected format or type conversions
        logging.warning(f"Skipping malformed row (type conversion error): {line} | Error: {e}")
        return []

def run():
    """Defines and runs the Beam pipeline."""
    # --- PIPELINE OPTIONS ---
    options = PipelineOptions(
        runner='DataflowRunner',
        project=PROJECT_ID,
        job_name=f'gcs-to-bq-store-info-team{TEAM_NUMBER.lower()}',
        staging_location=STAGING_LOCATION,
        temp_location=TEMP_LOCATION,
        region=REGION,
        # Add save_main_session to make defined functions available to workers
        save_main_session=True
    )

    # BigQuery table schema
    table_schema = "store_nbr:INTEGER,city:STRING,state:STRING,type:STRING,cluster:INTEGER"

    # Log pipeline configuration before running
    logging.info(f"Starting Dataflow job: {options.get_all_options()['job_name']}")
    logging.info(f"Reading from: {INPUT_FILE}")
    logging.info(f"Writing to BigQuery table: {TABLE_SPEC}")
    logging.info(f"BigQuery schema: {table_schema}")

    # Define the pipeline
    with beam.Pipeline(options=options) as p:
        (
            p
            | 'ReadStoreData' >> beam.io.ReadFromText(INPUT_FILE, skip_header_lines=1) # Use skip_header_lines
            | 'ParseStoreCSV' >> beam.FlatMap(parse_stores_csv)
            | 'WriteStoresToBigQuery' >> beam.io.WriteToBigQuery(
                table=TABLE_SPEC,
                schema=table_schema,
                create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
                write_disposition=beam.io.BigQueryDisposition.WRITE_TRUNCATE
            )
        )
    logging.info(f"Job '{options.get_all_options()['job_name']}' submitted successfully.")

if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    run()


Writing pipeline_stores.py


In [20]:
!python pipeline_stores.py

INFO:root:Starting Dataflow job: gcs-to-bq-store-info-teamdn5
INFO:root:Reading from: gs://mgmt599-justinrizzo-assignment2/kaggle-store-sales/stores.csv
INFO:root:Writing to BigQuery table: mgmt599-justinrizzo-lab2:store_sales_team_DN5.store_info
INFO:root:BigQuery schema: store_nbr:INTEGER,city:STRING,state:STRING,type:STRING,cluster:INTEGER
INFO:root:Runner defaulting to pickling library: cloudpickle.
INFO:apache_beam.runners.dataflow.dataflow_runner:Pipeline has additional dependencies to be installed in SDK worker container, consider using the SDK container image pre-building workflow to avoid repetitive installations. Learn more on https://cloud.google.com/dataflow/docs/guides/using-custom-containers#prebuild
INFO:apache_beam.runners.dataflow.internal.apiclient:Starting GCS upload to gs://mgmt599-justinrizzo-assignment2/staging/gcs-to-bq-store-info-teamdn5.1753592283.132273/submission_environment_dependencies.txt...
INFO:apache_beam.runners.dataflow.internal.apiclient:Completed GC

#Build the Option A: Predict Total Daily Sales model

In [22]:
from google.cloud import bigquery
import pandas as pd

# --- Configuration ---
PROJECT_ID = "mgmt599-justinrizzo-lab2"
DATASET_ID = "store_sales_team_DN5"

# Initialize the BigQuery client
client = bigquery.Client(project=PROJECT_ID)

# --- Step 1: Create the aggregated daily_sales table ---
print("Creating aggregated daily_sales table...")

query_create_table = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.daily_sales` AS
SELECT
  date,
  SUM(sales) as total_sales,
  COUNT(DISTINCT store_nbr) as stores_active,
  AVG(onpromotion) as avg_promo_items
FROM `{PROJECT_ID}.{DATASET_ID}.sales_data`
WHERE date < '2017-08-01'  -- Keep last 2 weeks for testing
GROUP BY date
"""

# Run the query
job_table = client.query(query_create_table)
job_table.result()  # Wait for the job to complete

print(f"Table `{job_table.destination.table_id}` created successfully.")


# --- Step 2: Create the ARIMA_PLUS forecasting model ---
print("\nTraining the sales forecast model...")

query_create_model = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.sales_forecast_model`
OPTIONS(
  model_type='ARIMA_PLUS',
  time_series_timestamp_col='date',
  time_series_data_col='total_sales'
) AS
SELECT date, total_sales
FROM `{PROJECT_ID}.{DATASET_ID}.daily_sales`
"""

# Run the query
job_model = client.query(query_create_model)
job_model.result()  # Wait for the model training to complete

print("Model `sales_forecast_model` trained successfully.")


# --- Step 3: Evaluate the model performance ---
print("\nEvaluating model performance...")

query_evaluate_model = f"""
SELECT * FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.sales_forecast_model`)
"""

# Run the query and get the results as a DataFrame
df_evaluation = client.query(query_evaluate_model).to_dataframe()

print("Model Evaluation Metrics:")
display(df_evaluation)


# --- (Optional) Step 4: Display the created daily_sales table ---
print("\nFirst 5 rows of the new daily_sales table:")
df_daily_sales = client.query(f"SELECT * FROM `{PROJECT_ID}.{DATASET_ID}.daily_sales` ORDER BY date LIMIT 5").to_dataframe()
display(df_daily_sales)


Creating aggregated daily_sales table...
Table `daily_sales` created successfully.

Training the sales forecast model...
Model `sales_forecast_model` trained successfully.

Evaluating model performance...
Model Evaluation Metrics:


non_seasonal_p  non_seasonal_d  non_seasonal_q  has_drift  log_likelihood  \
0                2               1               1       True   -20027.727338   
1                2               1               1      False   -20028.780393   
2                0               1               3       True   -20028.247396   
3                0               1               3      False   -20029.386084   
4                1               1               2       True   -20028.911869   
5                1               1               2      False   -20030.142094   
6                0               1               2       True   -20030.307645   
7                0               1               2      False   -20031.597063   
8                1               1               1       True   -20032.665872   
9                1               1               1      False   -20033.962482   
10               0               1               1      False   -20044.634556   
11               0               1               1       True   -20043.646469   
12               3               1               0      False   -20047.477046   
13               3               1               0       True   -20046.914225   
14               2               1               0      False   -20084.596506   
15               2               1               0       True   -20084.224420   
16               1               1               0      False   -20170.067442   
17               1               1               0       True   -20169.860700   
18               0               1               0      False   -20284.727347   
19               0               1               0       True   -20284.622505   

             AIC      variance  seasonal_periods  has_holiday_effect  \
0   40065.454676  1.484550e+09  [WEEKLY, YEARLY]               False   
1   40065.560786  1.486429e+09  [WEEKLY, YEARLY]               False   
2   40066.494791  1.485470e+09  [WEEKLY, YEARLY]               False   
3   40066.772168  1.487501e+09  [WEEKLY, YEARLY]               False   
4   40067.823737  1.486643e+09  [WEEKLY, YEARLY]               False   
5   40068.284189  1.488856e+09  [WEEKLY, YEARLY]               False   
6   40068.615290  1.489130e+09  [WEEKLY, YEARLY]               False   
7   40069.194126  1.491435e+09  [WEEKLY, YEARLY]               False   
8   40073.331744  1.493334e+09  [WEEKLY, YEARLY]               False   
9   40073.924965  1.495659e+09  [WEEKLY, YEARLY]               False   
10  40093.269112  1.514905e+09  [WEEKLY, YEARLY]               False   
11  40093.292938  1.513111e+09  [WEEKLY, YEARLY]               False   
12  40102.954092  1.520148e+09  [WEEKLY, YEARLY]               False   
13  40103.828450  1.519124e+09  [WEEKLY, YEARLY]               False   
14  40175.193013  1.589292e+09  [WEEKLY, YEARLY]               False   
15  40176.448839  1.588585e+09  [WEEKLY, YEARLY]               False   
16  40344.134885  1.760592e+09  [WEEKLY, YEARLY]               False   
17  40345.721400  1.760157e+09  [WEEKLY, YEARLY]               False   
18  40571.454695  2.019571e+09  [WEEKLY, YEARLY]               False   
19  40573.245009  2.019318e+09  [WEEKLY, YEARLY]               False   

    has_spikes_and_dips  has_step_changes  
0                  True              True  
1                  True              True  
2                  True              True  
3                  True              True  
4                  True              True  
5                  True              True  
6                  True              True  
7                  True              True  
8                  True              True  
9                  True              True  
10                 True              True  
11                 True              True  
12                 True              True  
13                 True              True  
14                 True              True  
15                 True              True  
16                 True              True


First 5 rows of the new daily_sales table:


date    total_sales  stores_active  avg_promo_items
0  2013-01-01    2406.618999             54              0.0
1  2013-01-02  493681.417944             54              0.0
2  2013-01-03  358985.231124             54              0.0
3  2013-01-04  349663.677093             54              0.0
4  2013-01-05  470635.121229             54              0.0

Model Predictions

In [23]:
from google.cloud import bigquery
import pandas as pd

# --- Configuration ---
PROJECT_ID = "mgmt599-justinrizzo-lab2"
DATASET_ID = "store_sales_team_DN5"

# Initialize the BigQuery client
client = bigquery.Client(project=PROJECT_ID)

# --- Step 1: Create the aggregated daily_sales table ---
print("Creating aggregated daily_sales table...")

query_create_table = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.daily_sales` AS
SELECT
  date,
  SUM(sales) as total_sales,
  COUNT(DISTINCT store_nbr) as stores_active,
  AVG(onpromotion) as avg_promo_items
FROM `{PROJECT_ID}.{DATASET_ID}.sales_data`
WHERE date < '2017-08-01'  -- Keep last 2 weeks for testing
GROUP BY date
"""

# Run the query
job_table = client.query(query_create_table)
job_table.result()  # Wait for the job to complete

print(f"Table `{job_table.destination.table_id}` created successfully.")


# --- Step 2: Create the ARIMA_PLUS forecasting model ---
print("\nTraining the sales forecast model...")

query_create_model = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.sales_forecast_model`
OPTIONS(
  model_type='ARIMA_PLUS',
  time_series_timestamp_col='date',
  time_series_data_col='total_sales'
) AS
SELECT date, total_sales
FROM `{PROJECT_ID}.{DATASET_ID}.daily_sales`
"""

# Run the query
job_model = client.query(query_create_model)
job_model.result()  # Wait for the model training to complete

print("Model `sales_forecast_model` trained successfully.")


# --- Step 3: Evaluate the model performance ---
print("\nEvaluating model performance...")

query_evaluate_model = f"""
SELECT * FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.sales_forecast_model`)
"""

# Run the query and get the results as a DataFrame
df_evaluation = client.query(query_evaluate_model).to_dataframe()

print("Model Evaluation Metrics:")
display(df_evaluation)


# --- Step 4: Make future sales predictions ---
print("\nMaking future sales predictions for the next 14 days...")

query_forecast_model = f"""
SELECT * FROM ML.FORECAST(MODEL `{PROJECT_ID}.{DATASET_ID}.sales_forecast_model`,
                          STRUCT(14 AS horizon))
"""

# Run the forecast query and get results as a DataFrame
df_forecast = client.query(query_forecast_model).to_dataframe()

print("Forecasted Sales:")
display(df_forecast)


# --- (Optional) Step 5: Display the created daily_sales table ---
print("\nFirst 5 rows of the new daily_sales table:")
df_daily_sales = client.query(f"SELECT * FROM `{PROJECT_ID}.{DATASET_ID}.daily_sales` ORDER BY date LIMIT 5").to_dataframe()
display(df_daily_sales)


Creating aggregated daily_sales table...
Table `daily_sales` created successfully.

Training the sales forecast model...
Model `sales_forecast_model` trained successfully.

Evaluating model performance...
Model Evaluation Metrics:


non_seasonal_p  non_seasonal_d  non_seasonal_q  has_drift  log_likelihood  \
0                2               1               1       True   -20027.727338   
1                2               1               1      False   -20028.780393   
2                0               1               3       True   -20028.247396   
3                0               1               3      False   -20029.386084   
4                1               1               2       True   -20028.911869   
5                1               1               2      False   -20030.142094   
6                0               1               2       True   -20030.307667   
7                0               1               2      False   -20031.597063   
8                1               1               1       True   -20032.665872   
9                1               1               1      False   -20033.962482   
10               0               1               1      False   -20044.634556   
11               0               1               1       True   -20043.646465   
12               3               1               0      False   -20047.477046   
13               3               1               0       True   -20046.914225   
14               2               1               0      False   -20084.596506   
15               2               1               0       True   -20084.224420   
16               1               1               0      False   -20170.067442   
17               1               1               0       True   -20169.860909   
18               0               1               0      False   -20284.727347   
19               0               1               0       True   -20284.622505   

             AIC      variance  seasonal_periods  has_holiday_effect  \
0   40065.454676  1.484550e+09  [WEEKLY, YEARLY]               False   
1   40065.560786  1.486429e+09  [WEEKLY, YEARLY]               False   
2   40066.494791  1.485470e+09  [WEEKLY, YEARLY]               False   
3   40066.772168  1.487501e+09  [WEEKLY, YEARLY]               False   
4   40067.823737  1.486643e+09  [WEEKLY, YEARLY]               False   
5   40068.284189  1.488856e+09  [WEEKLY, YEARLY]               False   
6   40068.615334  1.489129e+09  [WEEKLY, YEARLY]               False   
7   40069.194126  1.491435e+09  [WEEKLY, YEARLY]               False   
8   40073.331744  1.493334e+09  [WEEKLY, YEARLY]               False   
9   40073.924965  1.495659e+09  [WEEKLY, YEARLY]               False   
10  40093.269112  1.514905e+09  [WEEKLY, YEARLY]               False   
11  40093.292929  1.513111e+09  [WEEKLY, YEARLY]               False   
12  40102.954092  1.520148e+09  [WEEKLY, YEARLY]               False   
13  40103.828450  1.519124e+09  [WEEKLY, YEARLY]               False   
14  40175.193013  1.589292e+09  [WEEKLY, YEARLY]               False   
15  40176.448839  1.588585e+09  [WEEKLY, YEARLY]               False   
16  40344.134885  1.760592e+09  [WEEKLY, YEARLY]               False   
17  40345.721817  1.760156e+09  [WEEKLY, YEARLY]               False   
18  40571.454695  2.019571e+09  [WEEKLY, YEARLY]               False   
19  40573.245009  2.019318e+09  [WEEKLY, YEARLY]               False   

    has_spikes_and_dips  has_step_changes  
0                  True              True  
1                  True              True  
2                  True              True  
3                  True              True  
4                  True              True  
5                  True              True  
6                  True              True  
7                  True              True  
8                  True              True  
9                  True              True  
10                 True              True  
11                 True              True  
12                 True              True  
13                 True              True  
14                 True              True  
15                 True              True  
16                 True              True


Making future sales predictions for the next 14 days...
Forecasted Sales:


forecast_timestamp  forecast_value  standard_error  \
0  2017-08-01 00:00:00+00:00    8.733159e+05    38529.864286   
1  2017-08-02 00:00:00+00:00    8.757972e+05    41850.341907   
2  2017-08-03 00:00:00+00:00    7.232605e+05    43046.816781   
3  2017-08-04 00:00:00+00:00    8.345128e+05    44670.516758   
4  2017-08-05 00:00:00+00:00    1.016688e+06    46465.512184   
5  2017-08-06 00:00:00+00:00    1.107566e+06    48160.776061   
6  2017-08-07 00:00:00+00:00    8.088756e+05    49771.712439   
7  2017-08-08 00:00:00+00:00    6.901314e+05    51333.093672   
8  2017-08-09 00:00:00+00:00    7.022819e+05    52851.191997   
9  2017-08-10 00:00:00+00:00    6.155805e+05    54327.037197   
10 2017-08-11 00:00:00+00:00    7.738430e+05    55763.556329   
11 2017-08-12 00:00:00+00:00    9.740451e+05    57163.946540   
12 2017-08-13 00:00:00+00:00    1.069241e+06    58530.866598   
13 2017-08-14 00:00:00+00:00    7.991378e+05    59866.590783   

    confidence_level  prediction_interval_lower_bound  \
0               0.95                     7.979338e+05   
1               0.95                     7.939188e+05   
2               0.95                     6.390412e+05   
3               0.95                     7.471168e+05   
4               0.95                     9.257798e+05   
5               0.95                     1.013341e+06   
6               0.95                     7.114994e+05   
7               0.95                     5.897004e+05   
8               0.95                     5.988807e+05   
9               0.95                     5.092919e+05   
10              0.95                     6.647439e+05   
11              0.95                     8.622062e+05   
12              0.95                     9.547281e+05   
13              0.95                     6.820113e+05   

    prediction_interval_upper_bound  confidence_interval_lower_bound  \
0                      9.486980e+05                     7.979338e+05   
1                      9.576757e+05                     7.939188e+05   
2                      8.074798e+05                     6.390412e+05   
3                      9.219088e+05                     7.471168e+05   
4                      1.107595e+06                     9.257798e+05   
5                      1.201791e+06                     1.013341e+06   
6                      9.062519e+05                     7.114994e+05   
7                      7.905625e+05                     5.897004e+05   
8                      8.056830e+05                     5.988807e+05   
9                      7.218691e+05                     5.092919e+05   
10                     8.829420e+05                     6.647439e+05   
11                     1.085884e+06                     8.622062e+05   
12                     1.183755e+06                     9.547281e+05   
13                     9.162643e+05                     6.820113e+05   

    confidence_interval_upper_bound  
0                      9.486980e+05  
1                      9.576757e+05  
2                      8.074798e+05  
3                      9.219088e+05  
4                      1.107595e+06  
5                      1.201791e+06  
6                      9.062519e+05  
7                      7.905625e+05  
8                      8.056830e+05  
9                      7.218691e+05  
10                     8.829420e+05  
11                     1.085884e+06  
12                     1.183755e+06  
13                     9.162643e+05


First 5 rows of the new daily_sales table:


date    total_sales  stores_active  avg_promo_items
0  2013-01-01    2406.618999             54              0.0
1  2013-01-02  493681.417944             54              0.0
2  2013-01-03  358985.231124             54              0.0
3  2013-01-04  349663.677093             54              0.0
4  2013-01-05  470635.121229             54              0.0

# Task
Build a regression model based on the evaluation metrics and time series model, write the results of the data to markdown.

## Prepare data for regression

### Subtask:
Select relevant features and potentially engineer new ones for the regression model.


**Reasoning**:
Select relevant columns, convert date to datetime, and extract time-based features for the regression model.



In [24]:
import pandas as pd

# Select relevant columns
df_features = df_daily_sales[['date', 'total_sales', 'stores_active', 'avg_promo_items']].copy()

# Convert 'date' to datetime objects
df_features['date'] = pd.to_datetime(df_features['date'])

# Extract time-based features
df_features['year'] = df_features['date'].dt.year
df_features['month'] = df_features['date'].dt.month
df_features['day'] = df_features['date'].dt.day
df_features['dayofweek'] = df_features['date'].dt.dayofweek
df_features['weekofyear'] = df_features['date'].dt.isocalendar().week.astype(int)

# Display the first few rows with new features
display(df_features.head())

date    total_sales  stores_active  avg_promo_items  year  month  day  \
0 2013-01-01    2406.618999             54              0.0  2013      1    1   
1 2013-01-02  493681.417944             54              0.0  2013      1    2   
2 2013-01-03  358985.231124             54              0.0  2013      1    3   
3 2013-01-04  349663.677093             54              0.0  2013      1    4   
4 2013-01-05  470635.121229             54              0.0  2013      1    5   

   dayofweek  weekofyear  
0          1           1  
1          2           1  
2          3           1  
3          4           1  
4          5           1

**Reasoning**:
Create lag features for 'total_sales' and define the target variable (y) and features (X) for the regression model.



In [25]:
# Create lag features for 'total_sales' (e.g., sales from the previous day)
df_features['sales_lag_1'] = df_features['total_sales'].shift(1)

# Drop rows with NaN created by lagging
df_features.dropna(inplace=True)

# Define the target variable (y)
y = df_features['total_sales']

# Define the features (X), excluding the original 'date' column and target
X = df_features.drop(columns=['date', 'total_sales'])

# Display the first few rows of X and y to verify
display(X.head())
display(y.head())

stores_active  avg_promo_items  year  month  day  dayofweek  weekofyear  \
1             54              0.0  2013      1    2          2           1   
2             54              0.0  2013      1    3          3           1   
3             54              0.0  2013      1    4          4           1   
4             54              0.0  2013      1    5          5           1   

     sales_lag_1  
1    2406.618999  
2  493681.417944  
3  358985.231124  
4  349663.677093

1    493681.417944
2    358985.231124
3    349663.677093
4    470635.121229
Name: total_sales, dtype: float64

## Build the regression model

### Subtask:
Choose a suitable regression model (e.g., Linear Regression, Ridge, Lasso) and train it on the prepared data.


**Reasoning**:
Import the LinearRegression model and train it on the prepared feature data.



In [26]:
from sklearn.linear_model import LinearRegression

# Instantiate the Linear Regression model
model_regression = LinearRegression()

# Fit the model to the training data
model_regression.fit(X, y)

print("Linear Regression model trained successfully.")

Linear Regression model trained successfully.


## Evaluate the regression model

### Subtask:
Assess the model's performance using appropriate metrics (e.g., R-squared, Mean Squared Error).


**Reasoning**:
Calculate and print the Mean Squared Error and R-squared score for the regression model's predictions.



In [27]:
from sklearn.metrics import mean_squared_error, r2_score

# Make predictions using the trained regression model
y_pred = model_regression.predict(X)

# Calculate Mean Squared Error
mse = mean_squared_error(y, y_pred)
print(f"Mean Squared Error (MSE): {mse}")

# Calculate R-squared score
r2 = r2_score(y, y_pred)
print(f"R-squared (R2) Score: {r2}")

Mean Squared Error (MSE): 1301660830.2681472
R-squared (R2) Score: 0.6872980040121832


## Present the results

### Subtask:
Summarize the regression model's performance and key findings in markdown format.


**Reasoning**:
Summarize the regression model's performance and key findings in markdown format as requested by the instructions.



In [28]:
print("## Regression Model Performance")
print("")
print("Based on the Linear Regression model trained on the daily sales data, the evaluation metrics are as follows:")
print("")
print(f"- **Mean Squared Error (MSE):** {mse:.2f}")
print(f"- **R-squared (R2) Score:** {r2:.4f}")
print("")
print("### Interpretation of Metrics")
print("")
print(f"- The **MSE** of {mse:.2f} represents the average squared difference between the actual daily sales and the predicted daily sales. A lower MSE indicates a better fit of the model to the data.")
print(f"- The **R2 Score** of {r2:.4f} indicates that approximately {r2*100:.2f}% of the variance in daily sales can be explained by the features used in the Linear Regression model. An R2 score closer to 1 suggests that the model explains a large proportion of the variance in the target variable.")
print("")
print("This evaluation provides insights into how well the Linear Regression model captures the patterns in the daily sales data based on the selected features, including time-based components and the sales from the previous day.")

## Regression Model Performance

Based on the Linear Regression model trained on the daily sales data, the evaluation metrics are as follows:

- **Mean Squared Error (MSE):** 1301660830.27
- **R-squared (R2) Score:** 0.6873

### Interpretation of Metrics

- The **MSE** of 1301660830.27 represents the average squared difference between the actual daily sales and the predicted daily sales. A lower MSE indicates a better fit of the model to the data.
- The **R2 Score** of 0.6873 indicates that approximately 68.73% of the variance in daily sales can be explained by the features used in the Linear Regression model. An R2 score closer to 1 suggests that the model explains a large proportion of the variance in the target variable.

This evaluation provides insights into how well the Linear Regression model captures the patterns in the daily sales data based on the selected features, including time-based components and the sales from the previous day.


## Summary:

### Data Analysis Key Findings

*   The Mean Squared Error (MSE) for the Linear Regression model is approximately 1.3 billion (\$1,301,660,830.27), which represents the average squared difference between actual and predicted daily sales.
*   The R-squared (R2) score is approximately 0.6873, indicating that the Linear Regression model explains about 68.73% of the variance in daily sales using the selected features.

### Insights or Next Steps

*   While the R2 score suggests a reasonable fit, the high MSE indicates that the model's predictions still have a significant average squared error. Further feature engineering or exploring alternative models might improve accuracy.
*   Investigate the residuals of the model to understand where the largest prediction errors occur and identify potential areas for model improvement.


# Task
Analyze the "train.csv" and "stores.csv" datasets to identify key patterns and drivers of sales. Present the findings in markdown format.

## Load the dataset

### Subtask:
Load the `train.csv` and `stores.csv` files into pandas DataFrames.


**Reasoning**:
I will import the pandas library and load the `train.csv` and `stores.csv` files from their GCS paths into pandas DataFrames. Then I will display the head of each to verify they loaded correctly.



In [29]:
import pandas as pd

# Define the GCS paths for the train.csv and stores.csv files
train_file_path = f"gs://{GCS_BUCKET}/kaggle-store-sales/train.csv"
stores_file_path = f"gs://{GCS_BUCKET}/kaggle-store-sales/stores.csv"

# Read the train.csv file into a pandas DataFrame
df_train = pd.read_csv(train_file_path)

# Read the stores.csv file into a pandas DataFrame
df_stores = pd.read_csv(stores_file_path)

# Display the first few rows of the df_train DataFrame
print("Train Data:")
display(df_train.head())

# Display the first few rows of the df_stores DataFrame
print("\nStores Data:")
display(df_stores.head())

Train Data:


id        date  store_nbr      family  sales  onpromotion
0   0  2013-01-01          1  AUTOMOTIVE    0.0            0
1   1  2013-01-01          1   BABY CARE    0.0            0
2   2  2013-01-01          1      BEAUTY    0.0            0
3   3  2013-01-01          1   BEVERAGES    0.0            0
4   4  2013-01-01          1       BOOKS    0.0            0


Stores Data:


store_nbr           city                           state type  cluster
0          1          Quito                       Pichincha    D       13
1          2          Quito                       Pichincha    D       13
2          3          Quito                       Pichincha    D        8
3          4          Quito                       Pichincha    D        9
4          5  Santo Domingo  Santo Domingo de los Tsachilas    D        4

## Data cleaning and preparation

### Subtask:
Handle missing values, convert data types, and merge the datasets if necessary for combined analysis.


**Reasoning**:
Check for missing values in both dataframes and convert the 'date' column to datetime objects.



In [30]:
# Check for missing values in df_train
print("Missing values in df_train:")
display(df_train.isnull().sum())

# Check for missing values in df_stores
print("\nMissing values in df_stores:")
display(df_stores.isnull().sum())

# Convert 'date' column in df_train to datetime objects
df_train['date'] = pd.to_datetime(df_train['date'])

print("\n'date' column in df_train converted to datetime.")
display(df_train.info())

Missing values in df_train:


id             0
date           0
store_nbr      0
family         0
sales          0
onpromotion    0
dtype: int64


Missing values in df_stores:


store_nbr    0
city         0
state        0
type         0
cluster      0
dtype: int64


'date' column in df_train converted to datetime.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000888 entries, 0 to 3000887
Data columns (total 6 columns):
 #   Column       Dtype         
---  ------       -----         
 0   id           int64         
 1   date         datetime64[ns]
 2   store_nbr    int64         
 3   family       object        
 4   sales        float64       
 5   onpromotion  int64         
dtypes: datetime64[ns](1), float64(1), int64(3), object(1)
memory usage: 137.4+ MB


None

**Reasoning**:
Merge the two dataframes on 'store_nbr' and display the head and info of the merged dataframe to verify the merge and structure.



In [31]:
# Merge df_train and df_stores on 'store_nbr'
df_merged = pd.merge(df_train, df_stores, on='store_nbr', how='left')

# Display the first few rows of the merged DataFrame
print("\nMerged DataFrame Head:")
display(df_merged.head())

# Display information about the merged DataFrame
print("\nMerged DataFrame Info:")
display(df_merged.info())


Merged DataFrame Head:


id       date  store_nbr      family  sales  onpromotion   city      state  \
0   0 2013-01-01          1  AUTOMOTIVE    0.0            0  Quito  Pichincha   
1   1 2013-01-01          1   BABY CARE    0.0            0  Quito  Pichincha   
2   2 2013-01-01          1      BEAUTY    0.0            0  Quito  Pichincha   
3   3 2013-01-01          1   BEVERAGES    0.0            0  Quito  Pichincha   
4   4 2013-01-01          1       BOOKS    0.0            0  Quito  Pichincha   

  type  cluster  
0    D       13  
1    D       13  
2    D       13  
3    D       13  
4    D       13


Merged DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000888 entries, 0 to 3000887
Data columns (total 10 columns):
 #   Column       Dtype         
---  ------       -----         
 0   id           int64         
 1   date         datetime64[ns]
 2   store_nbr    int64         
 3   family       object        
 4   sales        float64       
 5   onpromotion  int64         
 6   city         object        
 7   state        object        
 8   type         object        
 9   cluster      int64         
dtypes: datetime64[ns](1), float64(1), int64(4), object(4)
memory usage: 228.9+ MB


None

## Exploratory data analysis (eda)

### Subtask:
Analyze overall sales trends over time, investigate the impact of store characteristics (type, cluster, location) on sales, examine the relationship between promotions and sales, analyze sales patterns by product family, and explore the effect of date-related features (day of week, month, year) on sales.


**Reasoning**:
Analyze the sales data by grouping by date, store characteristics, promotions, product family, and date-related features to identify patterns.



In [32]:
# 1. Overall sales trends over time
daily_sales_trend = df_merged.groupby('date')['sales'].sum().reset_index()
print("Overall Daily Sales Trend:")
display(daily_sales_trend.head())
display(daily_sales_trend.tail())

# 2. Impact of store characteristics on sales (mean sales)
sales_by_store_type = df_merged.groupby('type')['sales'].mean().reset_index()
print("\nMean Sales by Store Type:")
display(sales_by_store_type.sort_values(by='sales', ascending=False))

sales_by_store_cluster = df_merged.groupby('cluster')['sales'].mean().reset_index()
print("\nMean Sales by Store Cluster:")
display(sales_by_store_cluster.sort_values(by='sales', ascending=False))

sales_by_store_city = df_merged.groupby('city')['sales'].mean().reset_index()
print("\nMean Sales by Store City:")
display(sales_by_store_city.sort_values(by='sales', ascending=False).head())
display(sales_by_store_city.sort_values(by='sales', ascending=True).head())

sales_by_store_state = df_merged.groupby('state')['sales'].mean().reset_index()
print("\nMean Sales by Store State:")
display(sales_by_store_state.sort_values(by='sales', ascending=False).head())
display(sales_by_store_state.sort_values(by='sales', ascending=True).head())


# 3. Relationship between promotions and sales
sales_by_promotion = df_merged.groupby('onpromotion')['sales'].mean().reset_index()
print("\nMean Sales by Promotion Status (onpromotion):")
display(sales_by_promotion.sort_values(by='sales', ascending=False))

# 4. Sales patterns by product family
sales_by_family = df_merged.groupby('family')['sales'].mean().reset_index()
print("\nMean Sales by Product Family:")
display(sales_by_family.sort_values(by='sales', ascending=False).head())
display(sales_by_family.sort_values(by='sales', ascending=True).head())


# 5. Effect of date-related features on sales (mean sales)
sales_by_year = df_merged.groupby(df_merged['date'].dt.year)['sales'].mean().reset_index()
sales_by_year.rename(columns={'date':'year'}, inplace=True)
print("\nMean Sales by Year:")
display(sales_by_year)

sales_by_month = df_merged.groupby(df_merged['date'].dt.month)['sales'].mean().reset_index()
sales_by_month.rename(columns={'date':'month'}, inplace=True)
print("\nMean Sales by Month:")
display(sales_by_month.sort_values(by='sales', ascending=False))

sales_by_dayofweek = df_merged.groupby(df_merged['date'].dt.dayofweek)['sales'].mean().reset_index()
sales_by_dayofweek.rename(columns={'date':'dayofweek'}, inplace=True)
print("\nMean Sales by Day of Week (0=Monday, 6=Sunday):")
display(sales_by_dayofweek.sort_values(by='sales', ascending=False))

Overall Daily Sales Trend:


date          sales
0 2013-01-01    2511.618999
1 2013-01-02  496092.417944
2 2013-01-03  361461.231124
3 2013-01-04  354459.677093
4 2013-01-05  477350.121229

date          sales
1679 2017-08-11  826373.722022
1680 2017-08-12  792630.535079
1681 2017-08-13  865639.677471
1682 2017-08-14  760922.406081
1683 2017-08-15  762661.935939


Mean Sales by Store Type:


type       sales
0    A  705.878743
3    D  350.979407
1    B  326.739714
4    E  269.121301
2    C  197.263301


Mean Sales by Store Cluster:


cluster        sales
4         5  1117.245254
13       14   708.227718
7         8   647.377856
10       11   603.507018
16       17   592.231511
5         6   342.661732
0         1   326.163967
11       12   324.461406
12       13   324.364108
3         4   296.572872
8         9   274.968339
1         2   260.170621
9        10   255.897546
15       16   238.598137
14       15   199.010111
2         3   194.416262
6         7   139.261863


Mean Sales by Store City:


city       sales
18    Quito  556.577090
2   Cayambe  509.709179
0    Ambato  362.632273
4     Daule  345.282181
12     Loja  339.379530

city       sales
16       Puyo   73.601845
14      Manta  125.173982
15     Playas  138.139340
10  Latacunga  190.583961
5   El Carmen  198.975799


Mean Sales by Store State:


state       sales
12   Pichincha  554.110358
15  Tungurahua  362.632273
8         Loja  339.379530
4       El Oro  300.392253
5   Esmeraldas  294.963151

state       sales
11      Pastaza   73.601845
10       Manabi  149.774588
3      Cotopaxi  190.583961
2    Chimborazo  204.921881
13  Santa Elena  205.517129


Mean Sales by Promotion Status (onpromotion):


onpromotion         sales
249          249  12377.413000
250          250  12231.239000
248          248  12169.133000
246          246  11490.971000
251          251  10599.315667
..           ...           ...
3              3    871.408092
2              2    662.925632
1              1    467.556532
0              0    158.246681
265          279      0.000000

[362 rows x 2 columns]


Mean Sales by Product Family:


family        sales
12  GROCERY I  3776.972100
3   BEVERAGES  2385.793151
30    PRODUCE  1349.352123
7    CLEANING  1072.416744
8       DAIRY   709.154889

family     sales
4             BOOKS  0.070797
1         BABY CARE  0.110528
17  HOME APPLIANCES  0.457476
14         HARDWARE  1.137833
23        MAGAZINES  2.929082


Mean Sales by Year:


year       sales
0  2013  216.479530
1  2014  322.939786
2  2015  371.357193
3  2016  443.790297
4  2017  480.124466


Mean Sales by Month:


month       sales
11     12  453.740370
10     11  375.681762
6       7  374.219111
9      10  362.407132
8       9  362.297418
5       6  353.597842
2       3  352.009294
0       1  341.921554
4       5  341.765820
3       4  339.199841
7       8  336.992535
1       2  320.928869


Mean Sales by Day of Week (0=Monday, 6=Sunday):


dayofweek       sales
6          6  463.085366
5          5  433.336472
0          0  346.544732
2          2  332.909401
4          4  325.238138
1          1  319.823843
3          3  283.540517

## Feature engineering

### Subtask:
Create new features that might be relevant for understanding sales drivers (e.g., lag features, moving averages, indicators for holidays or special events if available).


**Reasoning**:
Create lag and rolling mean features for sales, and display the updated dataframe head to show the new features.



In [33]:
# Create lag features for 'sales'
df_merged['sales_lag_1'] = df_merged.groupby(['store_nbr', 'family'])['sales'].shift(1)
df_merged['sales_lag_7'] = df_merged.groupby(['store_nbr', 'family'])['sales'].shift(7)
df_merged['sales_lag_365'] = df_merged.groupby(['store_nbr', 'family'])['sales'].shift(365)

# Calculate rolling mean of 'sales' (e.g., 7-day window)
df_merged['sales_rolling_mean_7d'] = df_merged.groupby(['store_nbr', 'family'])['sales'].transform(lambda x: x.rolling(window=7).mean())

# Display the first few rows with new features
display(df_merged.head())

id       date  store_nbr      family  sales  onpromotion   city      state  \
0   0 2013-01-01          1  AUTOMOTIVE    0.0            0  Quito  Pichincha   
1   1 2013-01-01          1   BABY CARE    0.0            0  Quito  Pichincha   
2   2 2013-01-01          1      BEAUTY    0.0            0  Quito  Pichincha   
3   3 2013-01-01          1   BEVERAGES    0.0            0  Quito  Pichincha   
4   4 2013-01-01          1       BOOKS    0.0            0  Quito  Pichincha   

  type  cluster  sales_lag_1  sales_lag_7  sales_lag_365  \
0    D       13          NaN          NaN            NaN   
1    D       13          NaN          NaN            NaN   
2    D       13          NaN          NaN            NaN   
3    D       13          NaN          NaN            NaN   
4    D       13          NaN          NaN            NaN   

   sales_rolling_mean_7d  
0                    NaN  
1                    NaN  
2                    NaN  
3                    NaN  
4                    NaN

**Reasoning**:
Check for missing values created by the new features, impute or drop them if necessary, and display the number of remaining missing values to confirm.



In [34]:
# Check for missing values after creating new features
print("Missing values after feature engineering:")
display(df_merged.isnull().sum())

# Fill missing values created by lags and rolling means with 0
# These NaNs are typically at the start of each group (store_nbr, family)
df_merged.fillna(0, inplace=True)

# Check for missing values again to confirm imputation
print("\nMissing values after imputation:")
display(df_merged.isnull().sum())

Missing values after feature engineering:


id                            0
date                          0
store_nbr                     0
family                        0
sales                         0
onpromotion                   0
city                          0
state                         0
type                          0
cluster                       0
sales_lag_1                1782
sales_lag_7               12474
sales_lag_365            650430
sales_rolling_mean_7d     10692
dtype: int64


Missing values after imputation:


id                       0
date                     0
store_nbr                0
family                   0
sales                    0
onpromotion              0
city                     0
state                    0
type                     0
cluster                  0
sales_lag_1              0
sales_lag_7              0
sales_lag_365            0
sales_rolling_mean_7d    0
dtype: int64

**Reasoning**:
Display the first few rows of the dataframe to show the newly engineered features and confirm that the missing values were filled.



In [35]:
# Display the first few rows of the DataFrame to show the newly engineered features
display(df_merged.head())

id       date  store_nbr      family  sales  onpromotion   city      state  \
0   0 2013-01-01          1  AUTOMOTIVE    0.0            0  Quito  Pichincha   
1   1 2013-01-01          1   BABY CARE    0.0            0  Quito  Pichincha   
2   2 2013-01-01          1      BEAUTY    0.0            0  Quito  Pichincha   
3   3 2013-01-01          1   BEVERAGES    0.0            0  Quito  Pichincha   
4   4 2013-01-01          1       BOOKS    0.0            0  Quito  Pichincha   

  type  cluster  sales_lag_1  sales_lag_7  sales_lag_365  \
0    D       13          0.0          0.0            0.0   
1    D       13          0.0          0.0            0.0   
2    D       13          0.0          0.0            0.0   
3    D       13          0.0          0.0            0.0   
4    D       13          0.0          0.0            0.0   

   sales_rolling_mean_7d  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0

## Identify key patterns and drivers

### Subtask:
Based on the exploratory data analysis and engineered features, identify the most significant factors influencing sales.


**Reasoning**:
Analyze the results from the EDA and the newly engineered features to synthesize the key drivers of sales and present the findings in markdown.



In [36]:
# Summarize the key findings from EDA and feature engineering
print("## Key Drivers of Sales")
print("")
print("Based on the exploratory data analysis and engineered features, the most significant factors influencing sales appear to be:")
print("")

# 1. Impact of store characteristics (from EDA)
print("### Store Characteristics")
print("")
print(f"- **Store Type:** Store Type 'A' consistently shows the highest mean sales ({sales_by_store_type[sales_by_store_type['type'] == 'A']['sales'].iloc[0]:.2f}), while Type 'C' has the lowest ({sales_by_store_type[sales_by_store_type['type'] == 'C']['sales'].iloc[0]:.2f}). This suggests store type is a major determinant of sales volume.")
print(f"- **Store Cluster:** Cluster '5' exhibits exceptionally high mean sales ({sales_by_store_cluster[sales_by_store_cluster['cluster'] == 5]['sales'].iloc[0]:.2f}), significantly higher than other clusters. Cluster '7' has the lowest mean sales ({sales_by_store_cluster[sales_by_store_cluster['cluster'] == 7]['sales'].iloc[0]:.2f}). Cluster is a strong indicator of sales performance.")
print(f"- **Location (City/State):** Geographical location plays a role, with cities like Quito ({sales_by_store_city.sort_values(by='sales', ascending=False).iloc[0]['city']}) and states like Pichincha ({sales_by_store_state.sort_values(by='sales', ascending=False).iloc[0]['state']}) showing higher mean sales than others.")
print("")

# 2. Promotional activity (from EDA)
print("### Promotional Activity")
print("")
print(f"- **On Promotion:** Products on promotion have substantially higher mean sales values compared to those not on promotion. The number of items on promotion is strongly correlated with increased sales, indicating promotions are a powerful sales driver.")
print("")

# 3. Product families (from EDA)
print("### Product Families")
print("")
print(f"- **Top Performers:** 'GROCERY I' ({sales_by_family.sort_values(by='sales', ascending=False).iloc[0]['sales']:.2f}) and 'BEVERAGES' ({sales_by_family.sort_values(by='sales', ascending=False).iloc[1]['sales']:.2f}) are the product families with the highest mean sales, contributing significantly to overall revenue.")
print(f"- **Lower Performers:** Families like 'BOOKS' ({sales_by_family.sort_values(by='sales', ascending=True).iloc[0]['sales']:.2f}) have much lower mean sales.")
print("")

# 4. Time-based factors (from EDA)
print("### Time-Based Factors")
print("")
print(f"- **Yearly Trend:** There's a clear upward trend in mean sales year over year, indicating overall market growth or increasing store performance.")
print(f"- **Monthly Seasonality:** December ({sales_by_month.sort_values(by='sales', ascending=False).iloc[0]['month']}) shows the highest mean sales, likely due to seasonal demand. February ({sales_by_month.sort_values(by='sales', ascending=True).iloc[0]['month']}) has the lowest mean sales.")
print(f"- **Weekly Pattern:** Weekends (Saturday and Sunday) consistently have higher mean sales than weekdays, reflecting consumer shopping habits.")
print("")

# 5. Lagged sales and rolling means (from Feature Engineering)
print("### Historical Sales Data")
print("")
print("- **Lagged Sales:** Previous sales figures (e.g., sales from the previous day, week, or year) are likely strong predictors of current sales, capturing inherent time-series patterns and trends.")
print("- **Rolling Mean:** The rolling mean of sales provides a smoothed view of recent performance, which can help capture short-term trends and reduce noise, making it a useful feature for forecasting.")

print("")
print("In summary, store type, cluster, promotional activity, product family, and temporal factors (year, month, day of week), along with historical sales data, are the most influential drivers of sales in this dataset.")

## Key Drivers of Sales

Based on the exploratory data analysis and engineered features, the most significant factors influencing sales appear to be:

### Store Characteristics

- **Store Type:** Store Type 'A' consistently shows the highest mean sales (705.88), while Type 'C' has the lowest (197.26). This suggests store type is a major determinant of sales volume.
- **Store Cluster:** Cluster '5' exhibits exceptionally high mean sales (1117.25), significantly higher than other clusters. Cluster '7' has the lowest mean sales (139.26). Cluster is a strong indicator of sales performance.
- **Location (City/State):** Geographical location plays a role, with cities like Quito (Quito) and states like Pichincha (Pichincha) showing higher mean sales than others.

### Promotional Activity

- **On Promotion:** Products on promotion have substantially higher mean sales values compared to those not on promotion. The number of items on promotion is strongly correlated with increased sales, indicat

## Present findings

### Subtask:
Summarize the key patterns and insights found during the analysis in markdown format.


**Reasoning**:
Summarize the key patterns and insights found during the analysis in markdown format as requested by the instructions.



In [37]:
# 1. Create a markdown section titled "Summary of Sales Drivers Analysis".
print("## Summary of Sales Drivers Analysis")
print("")

# 2. Within this section, create a subsection titled "Key Findings" and use bullet points
print("### Key Findings")
print("")
print("- **Store Characteristics:**")
print(f"  - **Store Type:** Store Type 'A' has the highest mean sales ({sales_by_store_type[sales_by_store_type['type'] == 'A']['sales'].iloc[0]:.2f}), while Type 'C' has the lowest ({sales_by_store_type[sales_by_store_type['type'] == 'C']['sales'].iloc[0]:.2f}).")
print(f"  - **Store Cluster:** Cluster '5' shows the highest mean sales ({sales_by_store_cluster[sales_by_store_cluster['cluster'] == 5]['sales'].iloc[0]:.2f}), significantly outperforming other clusters. Cluster '7' has the lowest mean sales ({sales_by_store_cluster[sales_by_store_cluster['cluster'] == 7]['sales'].iloc[0]:.2f}).")
print(f"  - **Location:** Cities like Quito ({sales_by_store_city.sort_values(by='sales', ascending=False).iloc[0]['city']}) and states like Pichincha ({sales_by_store_state.sort_values(by='sales', ascending=False).iloc[0]['state']}) have higher average sales.")
print("")
print("- **Promotional Activity:** Products on promotion (`onpromotion` > 0) have substantially higher mean sales compared to those not on promotion (`onpromotion` = 0). The number of items on promotion is strongly correlated with increased sales.")
print("")
print("- **Product Families:** 'GROCERY I' ({sales_by_family.sort_values(by='sales', ascending=False).iloc[0]['sales']:.2f}) and 'BEVERAGES' ({sales_by_family.sort_values(by='sales', ascending=False).iloc[1]['sales']:.2f}) are the top-selling product families.")
print("")
print("- **Time-Based Factors:**")
print(f"  - **Yearly Trend:** Mean sales show a clear upward trend from 2013 to 2017.")
print(f"  - **Monthly Seasonality:** December ({sales_by_month.sort_values(by='sales', ascending=False).iloc[0]['month']}) has the highest mean sales, while February ({sales_by_month.sort_values(by='sales', ascending=True).iloc[0]['month']}) has the lowest.")
print(f"  - **Weekly Pattern:** Weekends (Saturday and Sunday) consistently have higher mean sales than weekdays.")
print("")
print("- **Historical Sales Data:** Lagged sales (e.g., `sales_lag_1`, `sales_lag_7`, `sales_lag_365`) and the 7-day rolling mean (`sales_rolling_mean_7d`) are important features that capture the temporal dynamics and trends in sales.")
print("")

# 3. Create another subsection titled "Insights and Implications"
print("### Insights and Implications")
print("")
print("The analysis reveals that sales are influenced by a combination of static store attributes, dynamic promotional efforts, product category popularity, and significant temporal patterns. These insights have several potential implications:")
print("")
print("- **Forecasting:** The identified time-based patterns (yearly trend, monthly, and weekly seasonality) and historical sales data (lags, rolling means) are crucial features for building accurate sales forecasting models.")
print("- **Targeted Marketing & Promotions:** Understanding which store types, clusters, and product families drive the highest sales allows for more targeted marketing campaigns and promotional strategies to maximize impact.")
print("- **Operational Improvements:** Insights into store performance variations by type, cluster, and location can inform decisions regarding inventory management, staffing, and resource allocation.")
print("- **Further Analysis:** The strong correlation between promotions and sales warrants deeper investigation into the effectiveness of different types of promotions. Analyzing sales by specific store within high/low performing clusters could also provide actionable insights.")

## Summary of Sales Drivers Analysis

### Key Findings

- **Store Characteristics:**
  - **Store Type:** Store Type 'A' has the highest mean sales (705.88), while Type 'C' has the lowest (197.26).
  - **Store Cluster:** Cluster '5' shows the highest mean sales (1117.25), significantly outperforming other clusters. Cluster '7' has the lowest mean sales (139.26).
  - **Location:** Cities like Quito (Quito) and states like Pichincha (Pichincha) have higher average sales.

- **Promotional Activity:** Products on promotion (`onpromotion` > 0) have substantially higher mean sales compared to those not on promotion (`onpromotion` = 0). The number of items on promotion is strongly correlated with increased sales.

- **Product Families:** 'GROCERY I' ({sales_by_family.sort_values(by='sales', ascending=False).iloc[0]['sales']:.2f}) and 'BEVERAGES' ({sales_by_family.sort_values(by='sales', ascending=False).iloc[1]['sales']:.2f}) are the top-selling product families.

- **Time-Based Factors:**
  -

## Summary:

## Summary of Sales Drivers Analysis

### Key Findings

*   **Store Characteristics:** Store Type 'A' consistently has the highest mean sales (\$705.88), while Type 'C' has the lowest (\$197.26). Store Cluster '5' shows the highest mean sales (\$1117.25), significantly outperforming other clusters, with Cluster '7' having the lowest (\$139.26). Cities like Quito and states like Pichincha have higher average sales.
*   **Promotional Activity:** Products on promotion have substantially higher mean sales compared to those not on promotion.
*   **Product Families:** 'GROCERY I' and 'BEVERAGES' are the top-selling product families.
*   **Time-Based Factors:** Mean sales show a clear upward trend from 2013 to 2017. December has the highest mean sales, while February has the lowest. Weekends (Saturday and Sunday) consistently have higher mean sales than weekdays.
*   **Historical Sales Data:** Lagged sales and the 7-day rolling mean capture temporal dynamics and trends in sales.

### Insights and Implications

*   The identified patterns and historical sales data are crucial features for building accurate sales forecasting models.
*   Understanding high-performing store types, clusters, and product families enables more targeted marketing and promotional strategies.


D - Discover

Based on our analysis of store sales data:

Our time-series model provides a strong fit to historical sales, automatically capturing weekly and yearly patterns, overall trends, and anomalies. The 14-day forecast confirms these patterns, predicting weekly sales peaks around \$1.1M on weekends and \$700k-\$800k on weekdays.

Key drivers of these sales include the performance of specific store types (Type 'A') and clusters (Cluster '5'), the popularity of product families like 'GROCERY I' and 'BEVERAGES', and the significant impact of promotions. Temporal factors, including weekend and December peaks, along with a consistent upward trend since 2013, also play a crucial role.

The main story these results tell about store sales is that they are predictable, driven by a combination of store-specific strengths, product mix, promotional activities, and clear seasonal and trending patterns.

---

**What did you discover at surface level?**

At a surface level, we discovered that store sales are driven by a predictable weekly and yearly cycle, are higher in certain store types and clusters, and are strongly influenced by promotions and top-selling product families.

I - Investigate

### Investigating the Drivers of Store Sales

The analysis highlights several key factors significantly influencing store sales: temporal patterns, store characteristics, and product/promotional effects. Understanding the "why" behind these patterns is crucial for strategic decision-making.

**Why do these specific factors influence sales so strongly?**

*   **Temporal Patterns (Weekly and Yearly Cycles):** The strong influence of weekly and yearly cycles is deeply rooted in fundamental customer behavior and societal rhythms. Weekends consistently show higher sales because this is when many people have time off from work and school, allowing for larger shopping trips, leisure spending, and stocking up for the week ahead. The yearly cycle, with a peak in December, is almost certainly driven by holiday shopping, seasonal demand for certain products, and potentially increased social activities leading to more purchases.
*   **Store Characteristics (Type 'A' and Cluster '5'):** Store characteristics like type and cluster are strong indicators of sales potential because they likely represent significant differences in store size, location, demographics of the surrounding area, and operational capabilities. Type 'A' stores and Cluster '5' stores may be larger, located in higher-traffic urban or suburban centers, serve denser or more affluent populations, or have better inventory management and customer service, all of which would directly contribute to higher sales volumes.
*   **Product & Promotion ('GROCERY I', 'BEVERAGES', Promotions):** The dominance of 'GROCERY I' and 'BEVERAGES' is likely due to their status as staple goods and high-frequency purchases. These are items most households need regularly. Promotions for these or other product families strongly influence sales because they directly impact price perception and value, often triggering impulse buys or encouraging customers to shift their purchases to when items are discounted. Promotions can also attract new customers or increase basket size for existing ones.

**What business mechanisms or customer behaviors explain these patterns?**

*   **Weekend vs. Weekday Sales:** Explained by customer availability for larger shopping trips and leisure spending. Businesses often align staffing and inventory with this predictable surge.
*   **Yearly Peak (December):** Driven by holiday gift-giving, seasonal food and beverage demands (e.g., for celebrations), and increased consumer spending during the festive period. Retailers strategically plan inventory and marketing around this peak.
*   **Store Type 'A' and Cluster '5':** As hypothesized, these likely correlate with factors like:
    *   **Location:** Prime spots with higher foot traffic or larger drawing areas.
    *   **Size and Assortment:** Larger stores capable of stocking a wider variety and deeper inventory, catering to more diverse needs.
    *   **Target Demographics:** Located in areas with higher purchasing power or larger populations.
    *   **Operational Efficiency:** Potentially better managed or more established stores.
*   **'GROCERY I' & 'BEVERAGES':** Reflect essential consumer needs. Weekly shopping habits often center around stocking up on these staples. Promotions on these items can influence brand choice or encourage bulk buying.
*   **Promotions:** Leverage psychological principles of value and urgency. Businesses use them to manage inventory, drive traffic, respond to competition, or boost sales of specific items. Customer behavior is driven by the desire for savings and perceived value. Promotions might align with typical weekly grocery runs or seasonal purchase timings.

**How do these factors interact?**

The interaction between these factors is complex and likely amplifies their individual effects:

*   **Promotions and Store Type/Cluster:** A promotion on a popular item like 'BEVERAGES' is likely to have a significantly larger impact in a high-traffic, high-volume Type 'A' or Cluster '5' store than in a smaller, lower-traffic Type 'C' store. This is because the potential customer base is larger, and the store might be better equipped to handle increased demand and display the promotion effectively.
*   **Temporal Patterns and Product/Promotion:** Promotions timed during peak seasons (like December) or peak shopping days (weekends) are likely to yield the highest returns. A promotion on 'GROCERY I' on a Saturday, when many customers are doing their main weekly shop, would likely be more impactful than the same promotion on a weekday.
*   **Store Location and Temporal Patterns:** Stores in areas with seasonal tourism might see different temporal patterns than those in residential areas. Stores in business districts might have stronger weekday sales patterns.

**What deeper patterns emerged?**

Beyond the surface-level observations, the deeper patterns that emerged are:

1.  **Reinforced Predictability:** The consistent weekly and yearly cycles, combined with the model's ability to capture anomalies, underscore a high degree of predictability in overall sales patterns. This suggests that while daily sales fluctuate, the underlying drivers create reliable rhythms.
2.  **Interconnected Influence:** Sales are not driven by single factors in isolation but by the complex interplay between store attributes, product demand, promotional strategies, and customer timing. High-performing stores amplify the impact of successful promotions and popular products, especially when timed with peak shopping periods.
3.  **Data Validation of Business Intuition:** The analysis validates common retail business intuition – that location matters, staples sell, promotions work, and timing is key – but quantifies these effects and integrates them into a predictive framework.

These deeper patterns emphasize the need for integrated strategies that consider *where* (Store Type/Cluster), *what* (Product Family), *when* (Temporal Patterns), and *how* (Promotions) to sell to maximize revenue and forecast accurately.

V- Validate

## Risk Analysis: Model Blind Spots

Based on the ARIMA\_PLUS model's reliance on historical data and its assumption that past patterns will continue, here's an analysis of its potential blind spots:

**What external events, not present in our historical data, could make the predictions wrong?**

The model is blind to unforeseen external events that could significantly impact sales patterns. These include:

*   **Unforeseen Economic Shifts:** A sudden recession, unexpected inflation spikes, or changes in consumer spending power not reflected in the historical period.
*   **Major Local Events:** Large-scale events like natural disasters (earthquakes, severe floods), significant public health crises (like a pandemic), or major local disruptions (e.g., prolonged strikes, civil unrest) that drastically alter shopping behavior or accessibility.
*   **Severe Weather:** Extreme or unusual weather patterns (e.g., blizzards, heatwaves, prolonged heavy rain) that prevent customers from reaching stores or impact demand for certain products, especially if these patterns are outside the historical range.
*   **Specific Competitor Actions:** Major strategic moves by competitors, such as opening a new large store nearby, aggressive price wars, or innovative promotional campaigns that draw customers away.
*   **Changes in Regulations or Policy:** New government regulations affecting retail operations, product availability, or consumer taxes.

**What are the key data limitations? How does the absence of this data create blind spots?**

The key data limitations explicitly mentioned are the absence of:

*   **External Economic Data:** Without data on factors like unemployment rates, GDP growth, or consumer confidence, the model cannot account for how broader economic health influences spending, creating a blind spot to macro-economic impacts.
*   **Weather Patterns:** Lacking specific weather data, the model cannot directly correlate sales with temperature, precipitation, or other weather conditions, missing a potential driver for certain product sales (e.g., seasonal items, beverages during heatwaves). This creates a blind spot for weather-driven demand fluctuations.
*   **Competitor Activities:** The absence of data on competitor pricing, promotions, or store openings means the model cannot anticipate how competitor actions might impact our sales, leaving it blind to competitive pressures and market share shifts.

The absence of these external data points means the model operates in a vacuum, assuming that the historical relationship between sales and the *available* features will hold, even if external conditions change dramatically.

**When would this model likely fail?**

This model would likely fail, or provide significantly inaccurate forecasts, in scenarios where:

*   There is a sudden, unprecedented shock to the economy or local environment (e.g., a pandemic lockdown, a major natural disaster).
*   A key competitor makes a significant strategic move that fundamentally changes the market landscape.
*   There are prolonged or extreme weather events outside the typical historical patterns.
*   Consumer behavior or preferences undergo a rapid and lasting shift not captured by the existing temporal or promotional data.
*   Significant changes occur within the stores themselves (e.g., major renovations, changes in management, stock issues) that are not reflected in the input features.

In essence, the model is best suited for forecasting under stable, predictable conditions that resemble the historical data. Any significant deviation from these past patterns, driven by external or uncaptured internal factors, increases the likelihood of forecast failure.

---

**What are the model's blind spots?**

The model's blind spots are external economic conditions, specific weather events, competitor actions, and any other unforeseen events or uncaptured internal changes that could disrupt historical sales patterns.

E - Extend

## Business Strategy Action Plan

Based on the key insight that sales are highly predictable and driven by specific factors, and the main risk that our forecasting model has blind spots to external events, here is a detailed action plan for store managers and regional directors:

### Next Week (Tactical Actions)

*   **Staffing Adjustment for Weekend Peak:**
    *   **Action:** Increase staffing levels, particularly in high-traffic areas and checkout counters, on Saturday and Sunday to manage the predicted ~$1.1M sales peak and ensure efficient customer service. Prioritize flexible scheduling to cover peak hours effectively.
    *   **Key Metric:** **Average Customer Checkout Time** or **Sales per Employee Hour** during peak weekend hours.

*   **Promotions on 'GROCERY I' and 'BEVERAGES':**
    *   **Action:** Implement targeted promotions (e.g., discounts, buy-one-get-one offers, prominent in-store displays) for 'GROCERY I' and 'BEVERAGES' specifically on the upcoming weekend to capitalize on predicted high traffic and product family popularity. Ensure sufficient inventory of promoted items.
    *   **Key Metric:** **Sales Volume and Revenue** for 'GROCERY I' and 'BEVERAGES' on the weekend, compared to previous non-promotion weekends.

### Next Month (Operational Planning)

*   **Applying Lessons from High-Performing Stores:**
    *   **Action:** Conduct a thorough analysis of operational practices, merchandising strategies, staffing models, and local marketing efforts in high-performing Type 'A' stores and Cluster '5' stores. Develop a playbook or best practices guide based on these findings. Pilot implement these best practices in a selection of underperforming Type 'C' or 'D' stores.
    *   **Key Metric:** **Percentage Increase in Mean Sales** in pilot Type 'C' or 'D' stores compared to a control group.

*   **Inventory Adjustments for Seasonal Trends:**
    *   **Action:** Review historical sales data (leveraging the yearly cycle insight) to identify products or categories that typically see increased demand in the months leading up to December. Proactively adjust inventory levels for these seasonal items across all stores, ensuring sufficient stock to meet anticipated demand.
    *   **Key Metric:** **In-Stock Percentage** for key seasonal product categories and **Reduction in Stockouts** during the seasonal period.

### Long-Term Planning (Strategic Initiatives)

*   **Strategy for New Store Locations:**
    *   **Action:** Prioritize site selection for new store locations in areas that share demographic or geographical characteristics with successful Cluster '5' areas, where feasible. Develop a more detailed understanding of the factors contributing to Cluster '5's success beyond just location (e.g., community engagement, local partnerships).
    *   **Key Metric:** **Average Sales Performance** of new stores opened in locations aligned with Cluster '5' characteristics, compared to stores opened in other types of locations.

*   **System to Monitor External Risks:**
    *   **Action:** Establish a cross-functional team responsible for monitoring key external indicators, including local economic news, competitor announcements (new stores, major promotions), and publicly available severe weather forecasts. Integrate relevant external data feeds (e.g., economic indicators, weather APIs) into our data analysis platform. Develop alert systems for significant changes in these external factors that could impact the sales forecast.
    *   **Key Metric:** **Number of Relevant External Risks Identified and Communicated** to relevant stakeholders before they significantly impact sales, and **Forecast Accuracy** during periods of external disruption.

---

**What should the business DO?**

The business should **leverage the predictable nature of sales** by tactically adjusting staffing and running targeted promotions on key products during peak times (weekends). It should **learn from its successes** by applying best practices from high-performing stores to improve others. Strategically, it should **seek expansion in areas resembling successful clusters** and **build a system to actively monitor external risks** to improve forecast resilience and proactive decision-making.